File: 200-number-of-islands/200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */

// 
const numIslands = (grid) => {
    const ROWS = grid.length,
          COLS = grid[0].length;
    
    let islands = 0;
    
    const flood = (r, c) => {
        if (
            r < 0 || 
            c < 0 || 
            r >= ROWS || 
            c >= COLS || 
            grid[r][c] === '0'
        ) {
            return;
        }
        grid[r][c] = '0';
        flood(r+1,c);
        flood(r-1,c);
        flood(r,c+1);
        flood(r,c-1);
    }
    
    for (let r = 0; r < ROWS; r++) {
        for (let c = 0; c < COLS; c++) {
            if (grid[r][c] === '1') {
                islands++;
                flood(r, c);
            }
            
        }
    }
    
    return islands;
};

File: 200-number-of-islands/200-number-of-islands.rb
# @param {Character[][]} grid
# @return {Integer}
def num_islands(grid)
    dirs = [[1,0],[-1,0],[0,1],[0,-1]]
    islands = 0
    grid.each_with_index do |row,r|
        row.each_with_index do |x,c|
            if x == '1'
                islands += 1
                flood grid, dirs, r, c
            end
        end
    end
    islands
end

def flood(grid,dirs,r,c)
    if is_valid(grid,r,c) && grid[r][c] == '1'
        grid[r][c] = '0'
        dirs.each { |dr,dc| flood(grid,dirs,r+dr,c+dc)}
    end
end

def is_valid(grid,r,c)
    r >= 0 && c >= 0 && r < grid.size && c < grid[0].size 
end